File: sem4/MPP/mpp-proiect-repository-mariaruncan/ProjJavaFX/Model/src/main/java/app/model/Game.java
package app.model;

import com.fasterxml.jackson.annotation.JsonIgnore;

import java.io.Serializable;
import java.time.LocalDate;
import java.util.Objects;

public class Game extends Entity<Long> implements Serializable {
    private Team team1;
    private Team team2;
    private LocalDate date;
    private String description;
    private Integer totalNoOfTickets;
    private Integer soldTickets;
    private Float pricePerTicket;

    public Game(){
        this.team1 = null;
        this.team2 = null;
        this.date = null;
        this.description = "";
        this.totalNoOfTickets = 0;
        this.soldTickets = 0;
        this.pricePerTicket = 0F;
    }

    public Game(Team team1, Team team2, LocalDate date, String description, Integer totalNoOfTickets, Integer soldTickets, Float pricePerTicket) {
        this.team1 = team1;
        this.team2 = team2;
        this.date = date;
        this.description = description;
        this.totalNoOfTickets = totalNoOfTickets;
        this.soldTickets = soldTickets;
        this.pricePerTicket = pricePerTicket;
    }

    public LocalDate getDate() {
        return date;
    }

    public void setDate(LocalDate date) {
        this.date = date;
    }

    public Team getTeam1() {
        return team1;
    }

    public void setTeam1(Team team1) {
        this.team1 = team1;
    }

    public Team getTeam2() {
        return team2;
    }

    public void setTeam2(Team team2) {
        this.team2 = team2;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public Integer getTotalNoOfTickets() {
        return totalNoOfTickets;
    }

    public void setTotalNoOfTickets(Integer totalNoOfTickets) {
        this.totalNoOfTickets = totalNoOfTickets;
    }

    public Integer getSoldTickets() {
        return soldTickets;
    }

    public void setSoldTickets(Integer soldTickets) {
        this.soldTickets = soldTickets;
    }

    public Float getPricePerTicket() {
        return pricePerTicket;
    }

    public void setPricePerTicket(Float pricePerTicket) {
        this.pricePerTicket = pricePerTicket;
    }

    @JsonIgnore
    public String getAvailableTickets(){
        if(Objects.equals(totalNoOfTickets, soldTickets))
            return "SOLD OUT!";
        return String.valueOf(totalNoOfTickets - soldTickets);
    }

    @Override
    public String toString() {
        return "Game: " + super.getId() + " " + date + " " + team1.getName() + " vs. " + team2.getName() +
                "\n" + description + "\nTickets available: " + (totalNoOfTickets - soldTickets) +
                " for only " + pricePerTicket + "$ per ticket";
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Game game = (Game) o;
        return Objects.equals(team1, game.team1) && Objects.equals(team2, game.team2) && Objects.equals(date, game.date)
                && Objects.equals(description, game.description) && Objects.equals(totalNoOfTickets, game.totalNoOfTickets)
                && Objects.equals(soldTickets, game.soldTickets) && Objects.equals(pricePerTicket, game.pricePerTicket);
    }

    @Override
    public int hashCode() {
        return Objects.hash(team1, team2, date, description, totalNoOfTickets, soldTickets, pricePerTicket);
    }
}

File: sem4/MPP/mpp-proiect-repository-mariaruncan/ProjCSharp/Model/Game.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Model
{
    [Serializable]
    public class Game : Entity<long>
    {
        public Team Team1 { get; set; }
        public Team Team2 { get; set; }
        public DateTime Date { get; set; }
        public string Description { get; set; }
        public int TotalNoOfTickets { get; set; }
        public int SoldTickets { get; set; }
        public float PricePerTicket { get; set; }

        public Game(): base(0)
        {
            Team1 = null;
            Team2 = null;
            Date = DateTime.MinValue;
            Description = "";
            TotalNoOfTickets = 0;
            SoldTickets = 0;
            PricePerTicket = 0;
        }

        public Game(Team team1, Team team2, DateTime date, string description, int totalNoOfTickets, int soldTickets, float pricePerTicket): base(0)
        {
            Team1 = team1;
            Team2 = team2;
            Date = date;
            Description = description;
            TotalNoOfTickets = totalNoOfTickets;
            SoldTickets = soldTickets;
            PricePerTicket = pricePerTicket;
        }

        public override string ToString()
        {
            return "Game: " + base.Id + " " + Date + " " + Team1.Name + " vs. " + Team2.Name +
                "\n" + Description + "\nTickets available: " + (TotalNoOfTickets - SoldTickets) +
                " for only " + PricePerTicket + "$ per ticket";
        }

        public override bool Equals(object obj)
        {
            return obj is Game game &&
                   Id == game.Id &&
                   EqualityComparer<Team>.Default.Equals(Team1, game.Team1) &&
                   EqualityComparer<Team>.Default.Equals(Team2, game.Team2) &&
                   Description == game.Description &&
                   TotalNoOfTickets == game.TotalNoOfTickets &&
                   SoldTickets == game.SoldTickets &&
                   PricePerTicket == game.PricePerTicket;
        }

        public override int GetHashCode()
        {
            int hashCode = 948494407;
            hashCode = hashCode * -1521134295 + Id.GetHashCode();
            hashCode = hashCode * -1521134295 + EqualityComparer<Team>.Default.GetHashCode(Team1);
            hashCode = hashCode * -1521134295 + EqualityComparer<Team>.Default.GetHashCode(Team2);
            hashCode = hashCode * -1521134295 + EqualityComparer<string>.Default.GetHashCode(Description);
            hashCode = hashCode * -1521134295 + TotalNoOfTickets.GetHashCode();
            hashCode = hashCode * -1521134295 + SoldTickets.GetHashCode();
            hashCode = hashCode * -1521134295 + PricePerTicket.GetHashCode();
            return hashCode;
        }

        public string AvailableTickets()
        {
            if (TotalNoOfTickets == SoldTickets)
            {
                return "SOLD OUT!";
            }
            return (TotalNoOfTickets - SoldTickets).ToString();
        }
    }
}
